File: problems/[E]350.Intersection-of-Two-Arrays-II.kt
/**
 * Runtime: 184 ms, faster than 88.89% of Kotlin online submissions for Intersection of Two Arrays II.
 * Memory Usage: 36.4 MB, less than 67.36% of Kotlin online submissions for Intersection of Two Arrays II.
 */
class Solution {
    fun intersect(nums1: IntArray, nums2: IntArray): IntArray {
        nums1.sort()
        nums2.sort()
        return if (nums1.size > nums2.size) {
            getIntersection(nums2, nums1)
        } else {
            getIntersection(nums1, nums2)
        }
    }
    
    fun getIntersection(smallerArray: IntArray, biggerArray: IntArray): IntArray {
        var sameItemsCount = 0
        var biggerIndex = 0
        for (i in 0 until smallerArray.size) {
            val number = smallerArray[i]
            while (biggerIndex < biggerArray.size && biggerArray[biggerIndex] < number) { 
                biggerIndex++ 
            }
            
            if (biggerIndex == biggerArray.size) break
            
            if (number == biggerArray[biggerIndex]) {
                sameItemsCount++
                biggerIndex++
            }
        }
        val result = IntArray(sameItemsCount)
        biggerIndex = 0
        var insertIndex = 0
        for (i in 0 until smallerArray.size) {
            val number = smallerArray[i]
            while (biggerIndex < biggerArray.size && biggerArray[biggerIndex] < number) { 
                biggerIndex++ 
            }
            
            if (biggerIndex == biggerArray.size) break
            
            if (number == biggerArray[biggerIndex]) {
                result[insertIndex] = number
                insertIndex++
                biggerIndex++
            }
        }
        return result
    }
}

File: python/[E]350.Intersection-of-Two-Arrays-II.py
# Runtime: 44 ms, faster than 86.25% of Python3 online submissions for Intersection of Two Arrays II.
# Memory Usage: 14.3 MB, less than 90.41% of Python3 online submissions for Intersection of Two Arrays II.
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        intCount = {}
        for number in nums1:
            intCount[number] = intCount.get(number, 0) + 1
        result = []
        for number in nums2:
            if intCount.get(number, 0) > 0:
                intCount[number] = intCount[number] - 1
                result.append(number)
        return result
